core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -333,6 +346,181 @@ public long nextFetchOffset() {
+                            if (throwable != null) {
+                                break;
+                            }
+                        } else {
+                            // The in-flight batch is a full match hence change the state of the complete batch.
+                            throwable = archiveCompleteBatch(inFlightBatch, updatedStates, stateBatches).orElse(null);
+                            if (throwable != null) {
+                                break;
+                            }
+                        }
+                    }
+                    // If the transitions to archive is successful then persist state, complete the state transition
+                    // and update the cached state for start offset. Else rollback the state transition.
+                    rollbackOrProcessArchivedStates(throwable, updatedStates, stateBatches);
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+        }
+    }
+
+
+    // This function provides the logic to get the earliest valid offset for a topic partition.
+    private long offsetForEarliestTimestamp() {
+        // TODO: We need to know the isolation level from group configs, for now we are passing Option.empty() for isolationLevel
+        Option<FileRecords.TimestampAndOffset> timestampAndOffset = replicaManager.fetchOffsetForTimestamp(
+                topicIdPartition.topicPartition(), ListOffsetsRequest.EARLIEST_TIMESTAMP, Option.empty(),
+                Optional.empty(), true);
+        return timestampAndOffset.isEmpty() ? (long) 0 : timestampAndOffset.get().offset;
+    }
+
+    private Optional<Throwable> archivePerOffsetBatchRecords(InFlightBatch inFlightBatch,
+                                                             long startOffset,
+                                                             long endOffset,
+                                                             List<InFlightState> updatedStates,
+                                                             List<PersisterStateBatch> stateBatches) {
+        Throwable throwable = null;
+        lock.writeLock().lock();
+        try {
+            log.trace("Archiving offset tracked batch: {} for the share partition: {}-{}", inFlightBatch, groupId, topicIdPartition);
+            for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState().entrySet()) {
+                if (offsetState.getKey() < startOffset) {
+                    continue;
+                }
+                if (offsetState.getKey() > endOffset) {
+                    // No further offsets to process.
+                    break;
+                }
+                if (offsetState.getValue().state == RecordState.ARCHIVED) {
+                    continue;
+                }
+
+                InFlightState updateResult = offsetState.getValue().startStateTransition(RecordState.ARCHIVED, false,
```

Comment:
I suppose this might happen if the records are ACQUIRED. Otherwise, it should be fine. I wonder whether actually we should permit the transition from ACQUIRED to ARCHIVED, as long as we handle the eventual acknowledgement which should leave these records in ARCHIVED. That seems like a better plan to me.